### ISSUE TITLE
[BUG] Can not access elements of complex arrays on distributed cluster

### ISSUE THREAD

**Issue by anders0821** [mars-project/mars#476]
Get errors when access elements of complex arrays on distributed cluster.

**To Reproduce**
Python 3.7.3, Ubuntu 19.04
pymars 0.2.0b1, clone from github
numpy 1.16.4, scipy 1.3.0, protobuf 3.6.1
Error message: "pyarrow.lib.ArrowNotImplementedError: Numpy scalar type not recognized" and "TypeError: Unknown type to serialize: <class 'numpy.complex128'>"
Toy code to reproduce the error:
```
DTYPE = mt.complex128
# DTYPE = mt.float64

a = np.zeros([2], DTYPE)
a = a[0]
print(a)  # OK

a = mt.zeros([2], DTYPE)
a = a[0]
print(a.execute())  # OK

a = mt.zeros([2], DTYPE)
a = a[0]
sess = new_session('http://192.168.1.186:7777')
print(a.execute(sess))  # worker: pyarrow.lib.ArrowNotImplementedError: Numpy scalar type not recognized
sess.close()
del sess

a = np.zeros([2], DTYPE)
a[0] = 1.0
print(a)  # OK

a = mt.zeros([2], DTYPE)
a[0] = 1.0
print(a.execute())  # OK

a = mt.zeros([2], DTYPE)
a[0] = 1.0
sess = new_session('http://192.168.1.186:7777')
print(a.execute(sess))  # client: TypeError: Unknown type to serialize: <class 'numpy.complex128'>
sess.close()
del sess
```

**Expected behavior**
Output the same result of the local computing

**Additional context**
N/A

**Comment by chineking** [comment]
Thanks you for your feedback. 

Mars leverages [pyarrow](https://arrow.apache.org/docs/python/) which is the python binding for arrow to serialize the data including numpy and pandas etc.

From the error message I can tell that pyarrow does not support serializing complex dtype yet, I checked the [docs](https://arrow.apache.org/docs/python/api/datatypes.html#factory-functions), complex type is not in the list indeed.

There is a solution that the complex128 is consist of 2 float64 numbers, so we can split the complex64 array into two float64 arrays until arrow supports complex dtype. But I think a better way is to report this to arrow team, and ask them for the complex128 dtype support.

**Comment by anders0821** [comment]
Serializing is a fundamental component used for transferring data between processes and machines. To my surprise, my mars distributed computing application is heavily based on complex numbers, and it works well. It seems that most mars operators support the complex128 in distributed computing. Are you sure that the problem is not related to the element access (or array slicing...)?

**Comment by chineking** [comment]
Could you please paste the full error stack so that I can make sure what is the root cause?

**Comment by anders0821** [comment]
Thank you for your attention.
Here is the first error stack (pyarrow.lib.ArrowNotImplementedError: Numpy scalar type not recognized):
```
Unexpected exception occurred in InProcessCacheActor.dump_cache.<locals>._try_put_chunk. chunk_key='44cf2e5cb6698e02043fb3e296525363'
Traceback (most recent call last):
  File "/home/anders/.local/lib/python3.7/site-packages/mars/utils.py", line 308, in _wrapped
    return func(*args, **kwargs)
  File "/home/anders/.local/lib/python3.7/site-packages/mars/worker/calc.py", line 78, in _try_put_chunk
    _calc_result_cache[session_chunk_key])
  File "/home/anders/.local/lib/python3.7/site-packages/mars/worker/chunkstore.py", line 204, in put
    serialized = pyarrow.serialize(value, self._serialize_context)
  File "pyarrow/serialization.pxi", line 355, in pyarrow.lib.serialize
  File "pyarrow/error.pxi", line 89, in pyarrow.lib.check_status
pyarrow.lib.ArrowNotImplementedError: Numpy scalar type not recognized
Exception met in executing promise.
Traceback (most recent call last):
  File "/home/anders/.local/lib/python3.7/site-packages/mars/promise.py", line 93, in _wrapped
    result = func(*args, **kwargs)
  File "/home/anders/.local/lib/python3.7/site-packages/mars/utils.py", line 308, in _wrapped
    return func(*args, **kwargs)
  File "/home/anders/.local/lib/python3.7/site-packages/mars/worker/calc.py", line 78, in _try_put_chunk
    _calc_result_cache[session_chunk_key])
  File "/home/anders/.local/lib/python3.7/site-packages/mars/worker/chunkstore.py", line 204, in put
    serialized = pyarrow.serialize(value, self._serialize_context)
  File "pyarrow/serialization.pxi", line 355, in pyarrow.lib.serialize
  File "pyarrow/error.pxi", line 89, in pyarrow.lib.check_status
pyarrow.lib.ArrowNotImplementedError: Numpy scalar type not recognized
Unhandled exception in promise
Traceback (most recent call last):
  File "/home/anders/.local/lib/python3.7/site-packages/mars/promise.py", line 93, in _wrapped
    result = func(*args, **kwargs)
  File "/home/anders/.local/lib/python3.7/site-packages/mars/utils.py", line 308, in _wrapped
    return func(*args, **kwargs)
  File "/home/anders/.local/lib/python3.7/site-packages/mars/worker/calc.py", line 78, in _try_put_chunk
    _calc_result_cache[session_chunk_key])
  File "/home/anders/.local/lib/python3.7/site-packages/mars/worker/chunkstore.py", line 204, in put
    serialized = pyarrow.serialize(value, self._serialize_context)
  File "pyarrow/serialization.pxi", line 355, in pyarrow.lib.serialize
  File "pyarrow/error.pxi", line 89, in pyarrow.lib.check_status
pyarrow.lib.ArrowNotImplementedError: Numpy scalar type not recognized
Unhandled exception in promise
Traceback (most recent call last):
  File "/home/anders/.local/lib/python3.7/site-packages/mars/promise.py", line 93, in _wrapped
    result = func(*args, **kwargs)
  File "/home/anders/.local/lib/python3.7/site-packages/mars/utils.py", line 308, in _wrapped
    return func(*args, **kwargs)
  File "/home/anders/.local/lib/python3.7/site-packages/mars/worker/calc.py", line 78, in _try_put_chunk
    _calc_result_cache[session_chunk_key])
  File "/home/anders/.local/lib/python3.7/site-packages/mars/worker/chunkstore.py", line 204, in put
    serialized = pyarrow.serialize(value, self._serialize_context)
  File "pyarrow/serialization.pxi", line 355, in pyarrow.lib.serialize
  File "pyarrow/error.pxi", line 89, in pyarrow.lib.check_status
pyarrow.lib.ArrowNotImplementedError: Numpy scalar type not recognized
Unhandled exception in promise
Traceback (most recent call last):
  File "/home/anders/.local/lib/python3.7/site-packages/mars/promise.py", line 93, in _wrapped
    result = func(*args, **kwargs)
  File "/home/anders/.local/lib/python3.7/site-packages/mars/utils.py", line 308, in _wrapped
    return func(*args, **kwargs)
  File "/home/anders/.local/lib/python3.7/site-packages/mars/worker/calc.py", line 78, in _try_put_chunk
    _calc_result_cache[session_chunk_key])
  File "/home/anders/.local/lib/python3.7/site-packages/mars/worker/chunkstore.py", line 204, in put
    serialized = pyarrow.serialize(value, self._serialize_context)
  File "pyarrow/serialization.pxi", line 355, in pyarrow.lib.serialize
  File "pyarrow/error.pxi", line 89, in pyarrow.lib.check_status
pyarrow.lib.ArrowNotImplementedError: Numpy scalar type not recognized
Unexpected error occurred in executing graph 6892f166ac0f2ea907f7230156e1c7ef
Traceback (most recent call last):
  File "/home/anders/.local/lib/python3.7/site-packages/mars/promise.py", line 93, in _wrapped
    result = func(*args, **kwargs)
  File "/home/anders/.local/lib/python3.7/site-packages/mars/utils.py", line 308, in _wrapped
    return func(*args, **kwargs)
  File "/home/anders/.local/lib/python3.7/site-packages/mars/worker/calc.py", line 78, in _try_put_chunk
    _calc_result_cache[session_chunk_key])
  File "/home/anders/.local/lib/python3.7/site-packages/mars/worker/chunkstore.py", line 204, in put
    serialized = pyarrow.serialize(value, self._serialize_context)
  File "pyarrow/serialization.pxi", line 355, in pyarrow.lib.serialize
    check_status(SerializeObject(context, wrapped_value, &serialized.data))
  File "pyarrow/error.pxi", line 89, in pyarrow.lib.check_status
    raise ArrowNotImplementedError(message)
pyarrow.lib.ArrowNotImplementedError: Numpy scalar type not recognized
```
And the second one (TypeError: Unknown type to serialize: <class 'numpy.complex128'>):
```
Traceback (most recent call last):
  File "C:\Program Files\JetBrains\PyCharm 2018.3.5\helpers\pydev\pydevd.py", line 1741, in <module>
    main()
  File "C:\Program Files\JetBrains\PyCharm 2018.3.5\helpers\pydev\pydevd.py", line 1735, in main
    globals = debugger.run(setup['file'], None, None, is_module)
  File "C:\Program Files\JetBrains\PyCharm 2018.3.5\helpers\pydev\pydevd.py", line 1135, in run
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "C:\Program Files\JetBrains\PyCharm 2018.3.5\helpers\pydev\_pydev_imps\_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "D:/Documents/Partners/百度/v_sunwenyun-1-icode/fast_aer/2-benchmarks+basicaer-mars/ttt.py", line 153, in <module>
    print(a.execute(sess))  # client: TypeError: Unknown type to serialize: <class 'numpy.complex128'>
  File "C:\Users\anders\AppData\Local\Programs\Python\Python36\lib\site-packages\mars\tensor\core.py", line 377, in execute
    return self._data.execute(session, **kw)
  File "C:\Users\anders\AppData\Local\Programs\Python\Python36\lib\site-packages\mars\core.py", line 554, in execute
    return session.run(self, **kw)
  File "C:\Users\anders\AppData\Local\Programs\Python\Python36\lib\site-packages\mars\session.py", line 115, in run
    result = self._sess.run(*tileables, **kw)
  File "C:\Users\anders\AppData\Local\Programs\Python\Python36\lib\site-packages\mars\web\session.py", line 134, in run
    graph_json = graph.to_json()
  File "mars\graph.pyx", line 443, in mars.graph.DirectedGraph.to_json
  File "mars\serialize\core.pyx", line 634, in mars.serialize.core.Serializable.to_json
  File "mars\serialize\core.pyx", line 605, in mars.serialize.core.Serializable.serialize
  File "mars\serialize\core.pyx", line 723, in mars.serialize.core.Provider.serialize_model
  File "mars\serialize\core.pyx", line 150, in mars.serialize.core.Field.serialize
  File "mars\serialize\core.pyx", line 151, in mars.serialize.core.Field.serialize
  File "mars\serialize\jsonserializer.pyx", line 450, in mars.serialize.jsonserializer.JsonSerializeProvider.serialize_field
  File "mars\serialize\core.pyx", line 605, in mars.serialize.core.Serializable.serialize
  File "mars\serialize\core.pyx", line 723, in mars.serialize.core.Provider.serialize_model
  File "mars\serialize\core.pyx", line 150, in mars.serialize.core.Field.serialize
  File "mars\serialize\core.pyx", line 151, in mars.serialize.core.Field.serialize
  File "mars\serialize\jsonserializer.pyx", line 434, in mars.serialize.jsonserializer.JsonSerializeProvider.serialize_field
  File "mars\serialize\core.pyx", line 669, in mars.serialize.core.AttributeAsDict.serialize
  File "mars\serialize\core.pyx", line 728, in mars.serialize.core.Provider.serialize_attribute_as_dict
  File "mars\serialize\core.pyx", line 723, in mars.serialize.core.Provider.serialize_model
  File "mars\serialize\core.pyx", line 150, in mars.serialize.core.Field.serialize
  File "mars\serialize\core.pyx", line 151, in mars.serialize.core.Field.serialize
  File "mars\serialize\jsonserializer.pyx", line 461, in mars.serialize.jsonserializer.JsonSerializeProvider.serialize_field
  File "mars\serialize\jsonserializer.pyx", line 390, in mars.serialize.jsonserializer.JsonSerializeProvider._serialize_value
  File "mars\serialize\jsonserializer.pyx", line 386, in mars.serialize.jsonserializer.JsonSerializeProvider._serialize_untyped_value
TypeError: Unknown type to serialize: <class 'numpy.complex128'>
```

**Comment by chineking** [comment]
I did a little experiment, I found that pyarrow can serialize the ndarray of complex128 or complex64, but cannot recognize the scalar of this type.

```
In [38]: import pyarrow as pa                                                   

In [39]: pa.serialize(np.zeros(1, dtype=np.complex64))                          
Out[39]: <pyarrow.lib.SerializedPyObject at 0x123412ca8>

In [40]: pa.serialize(np.complex64(0))                                          
---------------------------------------------------------------------------
ArrowNotImplementedError                  Traceback (most recent call last)
<ipython-input-40-b79617f3edf2> in <module>
----> 1 pa.serialize(np.complex64(0))

~/Workspace/arrow/python/pyarrow/serialization.pxi in pyarrow.lib.serialize()

~/Workspace/arrow/python/pyarrow/error.pxi in pyarrow.lib.check_status()

ArrowNotImplementedError: Numpy scalar type not recognized
```

Thus I guess the getitem return a scalar that cannot be serialized by pyarrow.

**Comment by chineking** [comment]
Second error may be related to Mars. Mars serializers cannot deal with numpy type, I will open another issue later.

**Comment by chineking** [comment]
We will also open a JIRA for project arrow to track the first issue,  url will be pasted later.

**Comment by anders0821** [comment]
I agree with you. This bug is caused by **complex scalars serialization in pyarrow**. This can explain that the two exceptions are raised on the client and the worker, respectively. It is inconvenient to operate complex numbers in mars distributed computing until the fundamental serialization problem is fixed.

**Comment by wjsi** [comment]
The issue in pyarrow can be traced at https://issues.apache.org/jira/browse/ARROW-5703 .

**Comment by qinxuye** [comment]
Complex type is now supported for Mars serialize solved in #488 .

**Comment by wjsi** [comment]
Close this as no further comments. Feel free to reopen it when the problem occurs again.
